server/lib/oauth2-client/oauth2.getAccessToken.js

```diff
@@ -0,0 +1,100 @@
+ *  authorizationCode:  '25dbf26a066d947fca82a1f05ae5890f79a27687',
+ *  referer: 'http://localhost:1444/'
+ *  }
+ * );
+ */
+async function getAccessToken(serviceId, userId, authorizationCode, referer) {
+  // Find provider configuration
+  const tokenHost = await this.variable.getValue(OAUTH2.VARIABLE.TOKEN_HOST, serviceId);
+  const tokenPath = await this.variable.getValue(OAUTH2.VARIABLE.TOKEN_PATH, serviceId);
+  const authorizeHost = await this.variable.getValue(OAUTH2.VARIABLE.AUTHORIZE_HOST, serviceId);
+  const authorizePath = await this.variable.getValue(OAUTH2.VARIABLE.AUTHORIZE_PATH, serviceId);
+  const grantType = await this.variable.getValue(OAUTH2.VARIABLE.GRANT_TYPE, serviceId);
+  const additionalAccessTokenRequestAxtionParam = await this.variable.getValue(
+    OAUTH2.VARIABLE.ADDITIONAL_ACCESS_TOKEN_REQUEST_ACTION_PARAM,
+    serviceId,
+  );
+  const redirectUriSuffix = await this.variable.getValue(OAUTH2.VARIABLE.REDIRECT_URI_SUFFIX, serviceId);
+
+  const clientId = await this.variable.getValue(OAUTH2.VARIABLE.CLIENT_ID, serviceId, userId);
+  const secret = await this.variable.getValue(OAUTH2.VARIABLE.CLIENT_SECRET, serviceId, userId);
+
+  // Init credentials based on integration name
+  const credentials = {
+    client: {
+      id: clientId,
+      secret,
+    },
+    auth: {
+      tokenHost,
+      tokenPath,
+      authorizeHost,
+      authorizePath,
+    },
+  };
+
+  // Build token access request
+  const tokenConfig = {
+    code: authorizationCode,
+    client_id: clientId,
+    client_secret: secret,
+    grant_type: grantType,
+    redirect_uri: `${referer}${redirectUriSuffix}`,
+  };
+  if (additionalAccessTokenRequestAxtionParam) {
+    tokenConfig.action = additionalAccessTokenRequestAxtionParam;
+  }
+
+  try {
+    const client = new AuthorizationCode(credentials);
+    const authResult = await client.getToken(tokenConfig, { json: true });
+
+    if (authResult.token) {
+      if (authResult.token.status && authResult.token.status !== 0) {
+        throw new Error500('Oauth2 get token response is not with status 0');
+      }
+
+      let jsonResult;
+      if (authResult.token.body) {
+        jsonResult = authResult.token.body;
+      } else {
+        jsonResult = authResult.token;
+      }
+      if (!jsonResult.expires_at && jsonResult.expires_in) {
+        const expiredAt = new Date();
+        expiredAt.setSeconds(expiredAt.getSeconds() + jsonResult.expires_in);
+        jsonResult.expires_at = expiredAt.toISOString();
+      }
+
+      // Save accessToken
+      await this.variable.setValue(OAUTH2.VARIABLE.ACCESS_TOKEN, JSON.stringify(jsonResult), serviceId, userId);
+    }
+
+    return authResult;
+  } catch (error) {
+    await this.variable.destroy(OAUTH2.VARIABLE.CLIENT_ID, serviceId, userId);
+    await this.variable.destroy(OAUTH2.VARIABLE.CLIENT_SECRET, serviceId, userId);
+    logger.error(error);
+    throw new Error500(error);
```

Comment:
same here, throw the real error, not 500

Reply:
I throw origin error